### DISCUSSION TITLE
Strip all metadata except the color profile?

### DISCUSSION THREAD

**Discussion by grishka** [libvips/libvips#2283] [Q&A] [answered]
I'd like to write a webp file while stripping metadata like EXIF but keeping the color profile so the colors always look accurate. I'm using libvips in a server application and the files I'm processing are uploaded by users, so it's important to not leak any sensitive data that might be in EXIF tags. I'm currently passing `strip=true` to `vips_image_write_to_file` but that strips everything, including the color profile.

**Comment by jcupitt** [comment, marked as answer]
Hi @grishka,

I would remove all metadata except the ICC profile. In Python, something like:

```python
# take a private copy of the image before we start modifying it 
image = image.copy()

for field in image.get_fields():
    if field != "icc-profile-data":
        image.remove(field)

image.write_to_file(filename)
```

See https://libvips.github.io/libvips/API/current/libvips-header.html#vips-image-get-fields

**Comment by o314** [comment]
Hello all.
I wonder if this is possible with the cli without calling the whole python stack.
FYI my codebase is in julia.

**Reply by grishka** [reply]
Here's how I did it in Java:
https://github.com/grishka/Smithereen/blob/bb0a659564d04b4f35f0ea62faa7073b4ecade4b/src/main/java/smithereen/storage/MediaStorageUtils.java#L32-L38

Relevant part of the JNA wrapper:
https://github.com/grishka/Smithereen/blob/bb0a659564d04b4f35f0ea62faa7073b4ecade4b/src/main/java/smithereen/libvips/VipsImage.java#L137-L169

edit: I didn't notice the CLI part, sorry

**Reply by jcupitt** [reply]
Sorry, the libvips CLI is pretty basic, it's not really supposed to be a full-fledged thing, the idea is to use a scripting language of some kind.

You could sort-of do it, for example:

```
$ vipsheader -f icc-profile-data k2.jpg
AAAMSExpbm8CEAAAbW50clJHQiBYWVogB84AAgAJAAYAMQAAYWNzcE1TRlQAAAAAS
... etc.
```

Gets the profile as a base64 string. You can pipe that to a decoder, eg.:

```
$ vipsheader -f icc-profile-data k2.jpg | base64 -d > x.icc
```

Now save the image with `strip`:

```
$ vips copy wtc.jpg x.jpg[strip]
```

Then reattach:

```
$ vips copy x.jpg y.jpg[profile=x.icc]
```

But this will not be very efficient heh.

**Reply by o314** [reply]
> Here's how I did it in Java:
I mention Julia you may have noticed, not Java. They are frequently confused.
Thanks anyway for taking time to answer .

**Reply by grishka** [reply]
No I just showed how to do it with the raw C API you can call from almost any language (via JNA in my case).

**Reply by o314** [reply]
Agree. Julia has very good c support too. This involves digging into C interop either coming from Java or Julia.
More work than simply wrapping the CLI.

**Reply by jcupitt** [reply]
libvips is designed to be simple to wrap, so it should only be a week or so of work. You can do a binding for the whole of libvips in only a few 100 lines of code.

**Reply by jcupitt** [reply]
I made a start at a julia binding: https://github.com/jcupitt/julia-vips

Though of course I don't know what I'm doing and it's probably all wrong haha. Any comments etc. very welcome.

**Reply by o314** [reply]
Great, thanks for the efforts ! I will have a look at it .

**Comment by o314** [comment]
First of all, thanks for the quick reply !
My post was not specifically targeting icc-profile but expressing a concern about how to handle metadata at the cli.
Currently, one can use the cmds :

```
vipsheader -f bar foo.jpg
vipsheader -a foo.jpg
```

I was searching for some forms of the kind :
```
# same ones, but opened to more options
vipsheader read -f bar foo.jpg
vipsheader read -a foo.jpg

# new - output to json
vipsheader read -f bar foo.jpg -o foo.json
vipsheader read -a foo.jpg -o foo.json

# new - intput from json
vipsheader write foo.jpg foo.json

# new - rm cmds
vipsheader rm -f bar foo.jpg
vipsheader rm -a foo.jpg
```
The last one may match strip.
Those are current practices around the mgt of metadata of multimedia.
Nothing really complex, An overwork of course.
I would have been happy to contribute but my cpp skills are not skilled at all

**Reply by jcupitt** [reply]
I think this is out of scope -- `vipsheader` does not write images, it only reads them.

The libvips CLI is intentionally very basic and does not try to be a complete solution (in the way that imagemagick is, for example). You end up implementing most of a scripting language in your CLI interface, and why do that when there are already so many excellent scripting languages? In my opinion, it's better to put our limited development effort into making language bindings.

I'm updating the php-vips binding at the moment, but perhaps I'll have a stab at Julia afterwards, it'd be interesting to learn something about it. A binding for R would be useful as well.

Someone else is very welcome to make a more complete CLI, of course.

**Comment by o314** [comment]
I was not talking about writing image but reading/writing **metadata** about image eg. width, height, xres, yres, date, etc ; mainly exif stuff, but filtered and cleaned between many variants like jpeg, tiff, webp, etc as done currently by libvips.

A json output will be nice for metadata to not have to deal with pattern outputted by vipsheader like

```
exif-ifd2-DateTimeOriginal: 2022:02:15 14:11:05 (2022:02:15 14:11:05, ASCII, 20 components, 20 bytes)
```

but more simply

```json
{
    "exif-ifd2":
    {
        "DateTimeOriginal": "2022-02-15T14-11-05"
    }
}
```

I agree that this would be absolutely out of scope if that was about image content.
It's only a question of format output , and of getting/putting/deleting into some dictionary of string.
Quite close IMHO to its current goal

*This may deserve a separate discussion*

**Reply by jcupitt** [reply]
I understand, but unfortunately you usually can't modify image metadata without reading and writing the whole image. VIPS format does support this, but it's very unusual.

I agree exif output should be improved, we should have an enhancement issue for that.
